File: src/zdock/ZDOCK.cpp
/**
 * Copyright (c) 2019, Arjan van der Velde, Weng Lab
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "ZDOCK.hpp"
#include "Exception.hpp"
#include <fstream>
#include <iostream>
#include <sstream>

namespace zdock {

ZDOCK::ZDOCK(const std::string &fn)
    : boxsize_(0), spacing_(0.0), isswitched_(false), ismzdock_(false),
      isfixed_(false), version_(0), symmetry_(0), filename_(fn) {
  read_();
}

void ZDOCK::read_() {
  std::vector<std::string> header;
  std::ifstream infile(filename_);
  std::string line;
  if (infile.is_open()) {
    predictions_.clear();
    bool headerdone = false;
    int linenum = 1;
    while (std::getline(infile, line)) {
      Prediction p;
      p.ismzdock = false;
      if (7 == std::sscanf(line.c_str(), "%lf\t%lf\t%lf\t%d\t%d\t%d\t%lf",
                           &p.rotation[0], &p.rotation[1], &p.rotation[2],
                           &p.translation[0], &p.translation[1],
                           &p.translation[2], &p.score)) {
        if (ismzdock_) {
          // M-ZDOCK established but 7-column prediction encountered
          throw ZDOCKInvalidFormat(filename_,
                                   "Invalid M-ZDOCK prediction (line " +
                                       std::to_string(linenum) + ")");
        }
        headerdone = true;
        predictions_.push_back(p);
      } else if (5 == std::sscanf(line.c_str(), "%lf\t%lf\t%d\t%d\t%lf",
                                  &p.rotation[0], &p.rotation[1],
                                  &p.translation[0], &p.translation[1],
                                  &p.score)) {
        if (!ismzdock_ && predictions_.size() > 0) {
          // ZDOCK established but 5-column prediction encountered
          throw ZDOCKInvalidFormat(filename_,
                                   "Invalid ZDOCK prediction (line " +
                                       std::to_string(linenum) + ")");
        }
        p.rotation[2] = 0.0;
        p.translation[2] = 0.0;
        ismzdock_ = true;
        p.ismzdock = true;
        headerdone = true;
        predictions_.push_back(p);
      } else if (!headerdone) {
        header.push_back(line);
      } else {
        throw ZDOCKInvalidFormat(filename_, "Invalid prediction (line " +
                                                std::to_string(linenum) + ")");
      }
      linenum++;
    }
  }

  // figure out header
  if (!ismzdock_) {
    if (5 == header.size()) {
      int sw;
      version_ = 1; // new style has five header rows
      if (3 != std::sscanf(header[0].c_str(), "%d\t%lf\t%d", &boxsize_,
                           &spacing_, &sw)) {
        throw ZDOCKInvalidFormat(filename_, "ZDOCK header error");
      }
      isswitched_ = !!sw;
      isfixed_ = false;
    } else if (4 == header.size()) {
      version_ = 0; // old style has 4 header rows
      isfixed_ = true;
      isswitched_ = false;
      if (2 !=
          std::sscanf(header[0].c_str(), "%d\t%lf", &boxsize_, &spacing_)) {
        throw ZDOCKInvalidFormat(filename_, "ZDOCK header error");
      }
    } else {
      // ZDOCK _must_ have 4 or 5 header rows
      throw ZDOCKInvalidFormat(filename_, "ZDOCK header error");
    }
  } else {
    if (3 != header.size()) {
      // M-ZDOCK _must_ have 3 header rows
      throw ZDOCKInvalidFormat(filename_, "M-ZDOCK header error");
    }
    if (3 != std::sscanf(header[0].c_str(), "%d\t%lf\t%d", &boxsize_, &spacing_,
                         &symmetry_)) {
      throw ZDOCKInvalidFormat(filename_, "M-ZDOCK header error");
    }
    if (3 > symmetry_) {
      throw ZDOCKInvalidFormat(filename_,
                               "M-ZDOCK symmetry cannot be less than 3");
    }
  }

  // receptor
  if (!isfixed_) {
    receptor_.rotation[2] = 0.0; // be sure to set this
    int count =
        std::sscanf(header[1].c_str(), "%lf\t%lf\t%lf", &receptor_.rotation[0],
                    &receptor_.rotation[1], &receptor_.rotation[2]);
    if ((3 != count && !ismzdock_) || (count < 2 && ismzdock_)) {
      throw ZDOCKInvalidFormat(filename_, "Unable to obtain receptor initial rotation");
    }
  }
  try {
    std::stringstream ss(header[(isswitched_ ? 4 : 3) - (!version_)]);
    ss >> receptor_.filename;
    ss >> receptor_.translation[0];
    ss >> receptor_.translation[1];
    ss >> receptor_.translation[2];
  } catch (const std::exception &e) {
    throw ZDOCKInvalidFormat(filename_,
                             "Unable to obtain receptor initial translation");
  }

  // ligand
  if (!ismzdock_) {
    if (3 != std::sscanf(header[(isfixed_ ? 1 : 2)].c_str(), "%lf\t%lf\t%lf",
                         &ligand_.rotation[0], &ligand_.rotation[1],
                         &ligand_.rotation[2])) {
      throw ZDOCKInvalidFormat(filename_,
                               "Unable to obtain receptor initial rotation");
    }
    try {
      std::stringstream ss(header[(isswitched_ ? 3 : 4) - (!version_)]);
      ss >> ligand_.filename;
      ss >> ligand_.translation[0];
      ss >> ligand_.translation[1];
      ss >> ligand_.translation[2];
    } catch (const std::exception &e) {
      throw ZDOCKInvalidFormat(filename_,
                               "Unable to obtain receptor initial translation");
    }
  }
}

// get m-zdock symmetry
int ZDOCK::symmetry() const {
  if (!ismzdock_) {
    throw ZDOCKUnsupported("symmetry() not supported for ZDOCK output");
  }
  return symmetry_;
}

// get ligand (unsupported for M-ZDOCK)
Structure &ZDOCK::ligand() {
  if (ismzdock_) {
    throw ZDOCKUnsupported("ligand() not supported for M-ZDOCK output");
  }
  return ligand_;
}

// get ligand (unsupported for M-ZDOCK)
const Structure &ZDOCK::ligand() const {
  if (ismzdock_) {
    throw ZDOCKUnsupported("ligand() not supported for M-ZDOCK output");
  }
  return ligand_;
}

// text representation of structure (in (m-)zdock format)
std::ostream &operator<<(std::ostream &os, const Structure &obj) {
  char s[256];
  snprintf(s, sizeof(s), "%s\t%.3f\t%.3f\t%.3f", obj.filename.c_str(),
           obj.translation[0], obj.translation[1], obj.translation[2]);
  os << s;
  return os;
}

// text representation of prediction (actual (m-)zdock.out format)
std::ostream &operator<<(std::ostream &os, const Prediction &obj) {
  char s[256];
  if (obj.ismzdock) {
    snprintf(s, sizeof(s), "%.6f\t%.6f\t%d\t%d\t%.2f", obj.rotation[0],
             obj.rotation[1], obj.translation[0], obj.translation[1],
             obj.score);
  } else {
    snprintf(s, sizeof(s), "%.6f\t%.6f\t%.6f\t%d\t%d\t%d\t%.3f",
             obj.rotation[0], obj.rotation[1], obj.rotation[2],
             obj.translation[0], obj.translation[1], obj.translation[2],
             obj.score);
  }
  os << s;
  return os;
}

// text representation of m-)zdock
std::ostream &operator<<(std::ostream &os, const ZDOCK &obj) {
  char s[256];
  if (obj.ismzdock_) {
    // M-ZDOCK header 1, 2
    snprintf(s, sizeof(s), "%d\t%.1f\t%d\n%.6f\t%.6f\t%.6f\n", obj.boxsize_,
             obj.spacing_, obj.symmetry_, obj.receptor_.rotation[0],
             obj.receptor_.rotation[1], obj.receptor_.rotation[2]);
  } else if (obj.isfixed_) {
    // ZDOCK (fixed) header 1, 2
    snprintf(s, sizeof(s), "%d\t%.1f\n%.6f\t%.6f\t%.6f\n", obj.boxsize_,
             obj.spacing_, obj.ligand_.rotation[0], obj.ligand_.rotation[1],
             obj.ligand_.rotation[2]);
  } else {
    // ZDOCK (new format / not fixed) header 1, 2
    snprintf(s, sizeof(s), "%d\t%.1f\t%d\n%.6f\t%.6f\t%.6f\n%.6f\t%.6f\t%.6f\n",
             obj.boxsize_, obj.spacing_, obj.isswitched_,
             obj.receptor_.rotation[0], obj.receptor_.rotation[1],
             obj.receptor_.rotation[2], obj.ligand_.rotation[0],
             obj.ligand_.rotation[1], obj.ligand_.rotation[2]);
  }
  os << s;
  // receptor / ligand filename and translation
  os << (obj.isswitched_ ? obj.ligand_ : obj.receptor_);
  if (!obj.ismzdock_) {
    // ligand / receptor filename and translation
    os << '\n' << (obj.isswitched_ ? obj.receptor_ : obj.ligand_);
  }
  // predictions
  for (const auto &x : obj.predictions_) {
    os << '\n' << x;
  }
  return os;
}

} // namespace zdock

File: python/zdock/ZDOCK.py
#!/usr/bin/env python3

# Copyright (c) 2019, Arjan van der Velde, Weng Lab
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice, this
#   list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import os
import re


class Structure(object):

    def __init__(self):
        self._filename = ""
        self._translation = (0.0, 0.0, 0.0)
        self._rotation = (0.0, 0.0, 0.0)

    @property
    def filename(self):
        return self._filename

    @property
    def translation(self):
        return self._translation

    @property
    def rotation(self):
        return self._rotation

    @filename.setter
    def filename(self, f):
        self._filename = f

    @translation.setter
    def translation(self, t):
        if len(t) > 2:
            self._translation = (float(t[0]), float(t[1]), float(t[2]))
        else:
            self._translation = (float(t[0]), float(t[1]), 0.0)

    @rotation.setter
    def rotation(self, r):
        if len(r) > 2:
            self._rotation = (float(r[0]), float(r[1]), float(r[2]))
        else:
            self._rotation = (float(r[0]), float(r[1]), 0.0)

    def __repr__(self):
        return "%s\t%.3f\t%.3f\t%.3f" % \
               (self.filename, self.translation[0], self.translation[1],
                self.translation[2])


class Prediction(object):

    def __init__(self):
        self._filename = ""
        self._translation = (0, 0, 0)
        self._rotation = (0.0, 0.0, 0.0)
        self._score = 0.0
        self._ismzdock = False

    @property
    def translation(self):
        return self._translation

    @property
    def rotation(self):
        return self._rotation

    @property
    def ismzdock(self):
        return self._ismzdock

    @property
    def score(self):
        return self._score

    @translation.setter
    def translation(self, t):
        if len(t) > 2:
            self._translation = (int(t[0]), int(t[1]), int(t[2]))
        else:
            self._translation = (int(t[0]), int(t[1]), 0)

    @rotation.setter
    def rotation(self, r):
        if len(r) > 2:
            self._rotation = (float(r[0]), float(r[1]), float(r[2]))
        else:
            self._rotation = (float(r[0]), float(r[1]), 0.0)

    @ismzdock.setter
    def ismzdock(self, m):
        self._ismzdock = (m and m)  # make it a bool

    @score.setter
    def score(self, s):
        self._score = float(s)

    def __repr__(self):
        if self.ismzdock:
            return "%.6f\t%.6f\t%d\t%d\t%.2f" % \
                   (self.rotation[0], self.rotation[1], self.translation[0],
                    self.translation[1], self.score)
        else:
            return "%.6f\t%.6f\t%.6f\t%d\t%d\t%d\t%.3f" % \
                   (self.rotation[0], self.rotation[1], self.rotation[2],
                    self.translation[0], self.translation[1],
                    self.translation[2], self.score)


class ZDOCK(object):

    def __init__(self, filename, n=None):
        self._receptor = Structure()
        self._ligand = Structure()
        self._predictions = []
        self._boxsize = 0
        self._spacing = 0.0
        self._isswitched = False
        self._ismzdock = False
        self._isfixed = False
        self._version = 0
        self._symmetry = 0
        self._filename = os.path.expanduser(os.path.normpath(filename))
        self._read(n)

    @property
    def receptor(self):
        return self._receptor

    @property
    def ligand(self):
        if self.ismzdock:
            raise Exception("ligand() not supported for M-ZDOCK")
        return self._ligand

    @property
    def npredictions(self):
        return len(self._predictions)

    @property
    def predictions(self):
        return self._predictions

    @property
    def ismzdock(self):
        return self._ismzdock

    @property
    def isfixed(self):
        return self._isfixed

    @property
    def symmetry(self):
        if not self.ismzdock:
            raise Exception("symmetry() not supported for M-ZDOCK")
        return self._symmetry

    @property
    def boxsize(self):
        return self._boxsize

    @property
    def filename(self):
        return self._filename

    @property
    def version(self):
        return self._version

    @property
    def spacing(self):
        return self._spacing

    @property
    def lines(self):
        return self._lines()

    @property
    def isswitched(self):
        return self._isswitched

    @ismzdock.setter
    def ismzdock(self, m):
        self._ismzdock = m and m  # make it bool

    @staticmethod
    def _procline(s):
        x = re.sub("^([^#]*).*$", "\\1", s).rstrip()
        if x:
            return x.split('\t')
        return []

    def _read(self, n=None):
        header = []
        with open(self._filename, 'r') as infile:
            self._predictions = []
            headerdone = False
            linenum = 1
            for line in infile:
                p = Prediction()
                p.ismzdock = False
                line = self._procline(line)
                if not line:
                    continue
                if 7 == len(line):
                    if self.ismzdock:
                        # M-ZDOCK established but 7-column prediction encountered
                        raise Exception("Invalid M-ZDOCK prediction (line " + str(linenum) + ")")
                    # ZDOCK prediction
                    (a, b, c, d, e, f, g) = \
                        [t(s) for t, s in zip((float, float, float, int, int, int, float), line)]
                    p.rotation = (a, b, c)
                    p.translation = (d, e, f)
                    p.score = g
                    self._predictions.append(p)
                    headerdone = True
                elif 5 == len(line):
                    if not self.ismzdock and self.npredictions > 0:
                        # ZDOCK established but 5-column prediction encountered
                        raise Exception("Invalid ZDOCK prediction (line " + str(linenum) + ")")
                    # M-ZDOCK prediction
                    (a, b, c, d, e) = [t(s) for t, s in zip((float, float, int, int, float), line)]
                    p.rotation = (a, b, 0.0)
                    p.translation = (c, d, 0)
                    p.score = e
                    p.ismzdock = True
                    self.ismzdock = True
                    self._predictions.append(p)
                    headerdone = True
                else:
                    if not headerdone:
                        header.append(line)
                    else:
                        raise Exception("Invalid prediction (line " + str(linenum) + ")")
                linenum += 1
                if n is not None and len(self._predictions) > n:
                    break

        # figure out header
        if not self.ismzdock:
            # ZDOCK
            if 5 == len(header):
                self._version = 1  # new style has 5 header rows
                try:
                    # string '0' to bool; str -> int -> bool
                    (self._boxsize, self._spacing, self._isswitched) = \
                        [t(s) for t, s in zip((int, float, lambda x: bool(int(x))), header[0])]
                    self._isfixed = False
                except Exception as e:
                    raise Exception("ZDOCK header error: " + str(e))
            elif 4 == len(header):
                self._version = 0  # old style has 4 header rows
                self._isfixed = True
                self._isswitched = False
                try:
                    (self._boxsize, self._spacing) = [t(s) for t, s in zip((int, float), header[0])]
                except Exception as e:
                    raise Exception("ZDOCK header error: " + str(e))
            else:
                raise Exception("ZDOCK header error; ZDOCK header must have 4 or 5 rows")
        else:
            # M-ZDOCK
            if 3 != len(header):
                raise Exception("M-ZDOCK header error; M-ZDOCK header must have 3 rows")
            try:
                (self._boxsize, self._spacing, self._symmetry) = \
                    [t(s) for t, s in zip((int, float, int), header[0])]
            except Exception as e:
                raise Exception("M-ZDOCK header error: " + str(e))
            if 3 > self.symmetry:
                raise Exception("M-ZDOCK symmetry cannot be less than 3")

        # receptor
        if not self.isfixed:
            try:
                if 3 == len(header[1]):
                    (a, b, c) = [t(s) for t, s in zip((float, float, float), header[1])]
                    self._receptor.rotation = (a, b, c)
                elif 2 == len(header[1]) and self.ismzdock:
                    (a, b) = [t(s) for t, s in zip((float, float), header[1])]
                    self._receptor.rotation = (a, b, 0.0)
                else:
                    raise Exception("Unable to obtain receptor initial rotation")
            except Exception as e:
                raise Exception("Unable to obtain receptor initial rotation: " + str(e))
        try:
            if self.isswitched:
                h = 4 - (not self.version)
            else:
                h = 3 - (not self.version)
            (a, b, c, d) = [t(s) for t, s in zip((str, float, float, float), header[h])]
            self._receptor.filename = a
            self._receptor.translation = (b, c, d)
        except Exception as e:
            raise Exception("Unable to obtain receptor initial translation: " + str(e))

        # ligand
        if not self.ismzdock:
            try:
                if self.isfixed:
                    h = 1
                else:
                    h = 2
                if 3 == len(header[h]):
                    (a, b, c) = [t(s) for t, s in zip((float, float, float), header[h])]
                    self._ligand.rotation = (a, b, c)
                else:
                    raise Exception("Unable to obtain ligand initial rotation")
            except Exception as e:
                raise Exception("Unable to obtain ligand initial rotation: " + str(e))
            try:
                if self.isswitched:
                    h = 3 - (not self.version)
                else:
                    h = 4 - (not self.version)
                (a, b, c, d) = [t(s) for t, s in zip((str, float, float, float), header[h])]
                self._ligand.filename = a
                self._ligand.translation = (b, c, d)
            except Exception as e:
                raise Exception("Unable to obtain ligand initial translation: " + str(e))

    def _lines(self):
        if self.ismzdock:
            yield "%d\t%.1f\t%d" % (self.boxsize, self.spacing, self.symmetry)
            yield "%.6f\t%.6f\t%.6f" % \
                  (self._receptor.rotation[0], self._receptor.rotation[1], self._receptor.rotation[2])
        elif self.isfixed:
            yield "%d\t%.1f" % (self.boxsize, self.spacing)
            yield "%.6f\t%.6f\t%.6f" % \
                  (self._ligand.rotation[0], self._ligand.rotation[1], self._ligand.rotation[2])
        else:
            yield "%d\t%.1f\t%d" % (self.boxsize, self.spacing, self.isswitched)
            yield "%.6f\t%.6f\t%.6f" % self._receptor.rotation
            yield "%.6f\t%.6f\t%.6f" % self._ligand.rotation
        if not self.ismzdock:
            if self.isswitched:
                yield str(self._ligand)
                yield str(self._receptor)
            else:
                yield str(self._receptor)
                yield str(self._ligand)
        else:
            yield str(self._receptor)
        for p in self._predictions:
            yield str(p)

    def __repr__(self):
        return '\n'.join([x for x in self._lines()])
